### PR TITLE
Internal Revamping of the Project

### PR DESCRIPTION
**Pull request by eksperimental**
Please run locally. Feedback is greatly appreciated
Ruby code is a bit rusty, please include your suggestions.

- Add jekyll-redirect-from plugin

- Move EN lessons to en/

- Interlanguage links no longer link to the current language being read (that is to the current page user one reading)

- lang property in `<html>` now uses the current language
  optionally a country_lang value can be set in _data/locales/  

- Move index.md to en/index.md
  By default if you access /, you will be redirected to the home of the english version in /en/

- Version Plugin has ben removed temporarily until the do all the changes, since changes in the plugin are needed due restructuring the project.

- Move translated texts from _config.yml to _data/locales/
  Translations are no longer stored in the config file.
  Each language has its own file under _data/locales/

- Set default layouts for pages in _config.yml
  Meaning we no longer need to define a layout in the front-matter.

- Add _config_dev.yml
  This was necessary because sometime directions will take me to elixirschool.com, while it has to be localhost.
  Additional config can be passed as:
  $ bundle exec jekyll s --config _config.yml,_config_dev.yml
  TODO: update README.md giving instructions

- Optimize _includes/head.html
  We no longer need to include title, as we are always printing site.title
  So we remove _includes/title.html

- Plugin to work with translations
  _plugins/languages.rb was added, and is the heart of this revamping.

  This plugin relieve the burden in Liquid by:

  - building a tree structure of all the chapters in the site
  - maps all the available translations for every chapter
  - maps interlang messages. basically how to say other language names in native and current language

  Additionally now it only lists available translations at the bottom of the sidebar,
  avoiding 404 errors, and the inconvenience of linking to a chapter that hasn't been translated in another language.

- Remove no longer needed front matter data from lessons: layout, category, order, lang
  This makes :title the only value to set. The rest is infered by the languages plugin.

- Move away from global variables and use page.data
  "Global" liquid variables are now defined by the languages plugin and stored under page.data

- Remove unnecessary fonts and reduce the size of the remaining ones
  - Add Icomoon font:
    We create a font with the font-awesome and entypo characters that we actually need (around 20)
  - Remove entypo font
  - Remove font-awesome (gem)

- Update gems

- Locales: rename some variables
  I have moved :title and :description
  under the key :project, since it was confusing I was confusing project title,
  with the page title.  
  :toc is renamed to :table_of_contents for clarity
  :home is added to :sections

- Temporarily manually redirect from / to /en/
  This is due to a bug in Jekyll that will redirect from /
  to //en/ instead of /en/

__UPDATES:__

- Rename chapters data file and menu variable. Pagination added to former menu var
   -  _data/chapters.yml has been renamed to _data/contents.yml and now
      it includes section :home, and file "home" as a section and chapter
    - Menu variable is renamed to Contents, which now includes home.
    - Pagination has been added to every item in content, indicating :previous and :next items
      by the array [section, chapter]

- Rename Languages plugin to ElixirSchool

-  Reimplement Version plugin as part of ElixirSchool plugin
    Version plugin has been moved from _plugin/version.rb to being parf of the main
    plugin ElixirSchool,
    It addes the "error" severity which will be returned when no version is defined in the
    chapter's frontmatter, or when any part of the version is higher in the translated page,
    or the version it is not in a valid format.
  
- `site.leaf` is added, which represent the current page from site.tree

### CODE REVIEW

**Comment by ybur-yug** [comment]
I'm going to read over all this tomorrow, but this is absolutely awesome @eksperimental. <3

**Comment by nscyclone** [comment]
Thanks, @eksperimental! That's astounding!
I'll try to find some time to play around with this ASAP.
Are we going to update the versioning plugin in this PR as well?

**Comment by eksperimental** [comment]
@nscyclone we could, or we could not. I would prefer someone else to do that if that's possible.

I have added a new commit that adds pagination information to the new contents hash, previously known as menu. So I guess we could use adapt your PR https://github.com/doomspork/elixir-school/pull/1100 to work with it once this is merged, so no need for a new pagination plugin.
Have a look and tell me what you think, or if you have any question, but Every item in the `site.contents[lang]` has `:previous` and `:next` which hold an array with `[section, chapter]`. Easy peasy.

File: jp/lessons/basics/sigils.md [hunk 1]
```diff
@@ -1,10 +1,6 @@
 ---
 version: 0.9.0
-layout: page
 title: Sigils
```
Comment (gemantzu) [inline jp/lessons/basics/sigils.md]: Can you clean this up? Windows new line to the rescue again.

Comment (eksperimental) [inline jp/lessons/basics/sigils.md]: do you mean the control characters at the end of the line?

**Comment by eksperimental** [comment]
@nscyclone I am working on migrating the version plugin to this one, so it will be covered in this PR.

File: jp/lessons/basics/sigils.md (hunk 1 above)
Comment (gemantzu) [inline jp/lessons/basics/sigils.md]: Yes yes, the new line windows character

**Comment by eksperimental** [comment]
@nscyclone version plugin has been moved to ElixirSchool plugin

**Comment by eksperimental** [comment]
I'm not planning on adding anything else to this PR, so please review, the only need to do it to update the documentation for this project (README.md, CONTRIBUTING.md) to reflect the new changes.

File: jp/lessons/basics/sigils.md (hunk 1 above)
Comment (eksperimental) [inline jp/lessons/basics/sigils.md]: @gemantzu done in a separate PR https://github.com/doomspork/elixir-school/pull/1122

**Comment by doomspork** [comment]
Thank you @eksperimental!  I've set aside time tomorrow to go through this and other outstanding PRs

**Comment by eksperimental** [comment]
Sorry guys, but it is just to let you all know that I will be available this upcoming week and after that I won't be available to work on this (and not only this PR but any other ones) for an indefinite amount of time. So if you want me to be the one fixing things and rebasing or answering questions, I'm glad to do it, but if that's not possible, someone else will have to take over it.

File: _plugins/elixir_school.rb [hunk 2]
```diff
@@ -137,15 +137,20 @@ def contents(site, lang)
         chapters.each_with_index do |chapter_name, index|
           if chapter = get_chapter(site, lang, section, chapter_name)
             chapter['chapter_number'] = index + 1
-            current = [section, chapter_name]
+            current = {
+              'section' => section,
+              'chapter' => chapter_name,
+              'url' => chapter['url'],
+              'title' => chapter['title']
+            }
```
Comment (eksperimental) [inline _plugins/elixir_school.rb]: I initially took this approach, but then i discarted by a two element array, because we are going to be duplicating information way too much. Since you have `site['tree']` there is no need to store `url` and `title` given `lang`, `chapter` and `section`.

**Comment by eksperimental** [comment]
@nscyclone could you remove the rebase commit, I will rebase it myself by the end of this week, without the need to introduce a commit that will add noise to the PR

**Comment by nscyclone** [comment]
@eksperimental, thanks for the review!
I've rebased the branch and ended up with only two commits of mine:
* the already mentioned sidebar sections' state fix;
* pagination implemented with use of `site.tree`.
Could you please verify if I understood your suggestion correctly?

File: _includes/sidebar.html [hunk 3]
```diff
@@ -12,15 +12,22 @@
     <nav class="sidebar-nav">
       {% for sections in site.contents[page.lang] %}
         {% assign _section = sections[0] | strip %}
+        {% if page.section == _section %}
+          {% assign opened = true %}
+        {% elsif page.section == "home" and forloop.index0 == 1 %}
+          {% assign opened = true %}
```
Comment (eksperimental) [inline _includes/sidebar.html]: I would prefer to go with `_section == 'basics'` instead of `forloop.index0 == 1`
The 2 clauses could fit in one, as in:
```
        {% if page.section == _section or (page.section == "home" and _section == "basics") %}
￼           {% assign opened = true %}
```

Comment (nscyclone) [inline _includes/sidebar.html]: Thanks, @eksperimental!
Concerning 2 condition with the same result, I really expected to see that in the review, because this way is pretty obvious and I did not want to go against the DRY principles. But for some reason, a complex condition like this one, did not work, and I even was stuck looking at an old Jekyll [issue](https://github.com/Shopify/liquid/issues/138) on Github about boolean algebra. Now everything seems to be fine, I don't even know what could possibly go wrong and I feel a little ashamed.

The choice of `forloop.index0 == 1` instead of `_section == 'basics'` was motivated by a desire to leave one less thing to worry about in case if the sections will be changed someday (although, it feels like this will likely never happen).

I've just updated the branch. Is there anything else I could help with?

Comment (eksperimental) [inline _includes/sidebar.html]: I didn't run that code. So please make sure it works.
the proper solution would be to create a config variable called "first_section", but that would be overkill IMO.

I will run your changes tomorrow as I haven't had time to do it yet (all my suggestions have been made by reading code not by viewing the live version :-P), and report back.

Comment (nscyclone) [inline _includes/sidebar.html]: I ran it locally before committing, seems to be working as intended.

Comment (nscyclone) [inline _includes/sidebar.html]: Actually we wouldn't need a new variable in config, because we already know the structure of the sections and chapters from [_contents.yml](https://github.com/doomspork/elixir-school/pull/1112/files#diff-71131372dfb994e3f094f9d9e2817101). So the index approach might make sense, but if we will finally decide to go this way, we should leave a comment to make clear what does the magic number mean.

Comment (eksperimental) [inline _includes/sidebar.html]: @nscyclone I have tested the pagination thing, it works great. There are two things I would like to comment.
1. It will be better if we can remove the underlining under the arrows. it looks weird.
2. When reading the first lesson, I think it makes more sense to use the word "Home" instead of the Title of the project, but if home hasn't been translated, then use the title of the project (which is Elixir School in English)

Comment (nscyclone) [inline _includes/sidebar.html]: @eksperimental, great idea. But I've a question: should we do something with the values of `sections.home` in the [locales](https://github.com/doomspork/elixir-school/pull/1112/files#diff-55286c7d0270bebf2261ad00b1964c8dR8)? The link to the index will be called 'Home' for every language if we leave those in the current state.

Comment (eksperimental) [inline _includes/sidebar.html]: @nscyclone well. I didn't remember I already added sections.home to every locale. So just use that one. Translators will have to translate a few things after this migration. I guess it is OK to rely on it, otherwise we will have to do , `if lang != 'en' and sections.home == 'Home', then use "project.title"` but that's would be a dirty hack and we don't know if some language borrow the English term Home  . So I would say let's rely on the translation of locale.

Comment (nscyclone) [inline _includes/sidebar.html]: Thanks, @eksperimental, that's exactly what I thought about.
I've pushed a small update addressing your recent feedback.

**Comment by eksperimental** [comment]
Rebased to master.
@nscyclone regarding your changes, I approve them, even though the option is not allowed here.
From my side we are good to go with this PR.

**Comment by doomspork** [comment]
Thank you @eksperimental / @nscyclone!!  I will pull this down and run through it this morning to make sure there aren't any issues and then we can merge this beast 💪

**Comment by doomspork** [comment]
This is wonderful!  Thanks again @eksperimental and @nscyclone for all of your hard work.

If you could rebase this branch and resolve the conflict I'll get this merge 👍

**Comment by eksperimental** [comment]
@doomspork done

**Comment by eksperimental** [comment]
@doomspork please merge, I will later write a file with what needs to be done. Which is basically fix the instructions in the README file reflecting the new changes. Also I would like to make a proposal for the /index page. Unfortunately I won't be able to contribute with code, maybe the README file but not the index page.
thank you guys

**Comment by eksperimental** [comment]
@doomspork also, please do not merge any other PR until this is merged. thank you

**Comment by doomspork** [comment]
@eksperimental I'll hold off but just as a FYI, each language has a translator with merge capability so it's not realistic to stop all PR merges on this repo.

**Comment by eksperimental** [comment]
Sorry but u asked for it to be rebased so it can be merged, i rebased but u merged another pr and this needs to be rebased again

**Comment by doomspork** [comment]
You're right @eksperimental, sorry about that.  I can take care of rebasing it if you'd rather, I don't mind 👍

**Comment by doomspork** [comment]
Rebased and pushed this to master.  Thanks again for all your hard work @eksperimental and @nscyclone!
